File: src/julia/modules/Parameters/src/load_general_configuration.jl
import Base: parse

function parse(::Type{String}, value::String)::String
	return value
end

function load_general_configuration!(gen_config_file_path::String, params::ParameterData)::Nothing
	# Create a dictionaty with fields and their types.
	param_names_types = Dict(
		[name => typ for (name, typ) in zip(fieldnames(ParameterData),
		ParameterData.types)]
	)

	param_given = Dict([name => false for name in keys(param_names_types)])

	lines = Array{String, 1}()
	try
		open(gen_config_file_path) do file
			lines = readlines(file)
		end
	catch err
		throw(LoadError(gen_config_file_path, 0,
			"cannot read '$gen_config_file_path'"))
	end

	if length(lines) == 0
		throw(LoadError(gen_config_file_path, 0,
			"cannot read '$gen_config_file_path'"))
	end

	for (line_number, line) in enumerate(lines)
		line = strip(line)
		if length(line) == 0 || line[1] == '#'
			continue
		end

		println("Loading line $line_number: $line")

		param_name = ""
		value = 0
		try
			param_name, value = split(line)

			field = Symbol(param_name)
			if field in fieldnames(ParameterData)
				data = params
			else
				throw(KeyError(""))
			end


			setfield!(data, field, parse(param_names_types[field], String(value)))
			param_given[field] = true
		catch err
			if isa(err, BoundsError)
				throw(
					LoadError(
						gen_config_file_path,
						line_number,
						"error line " *
						"$line_number of '$gen_config_file_path': " *
						"missing parameter or value",
					),
				)

			elseif isa(err, KeyError)
				throw(LoadError(gen_config_file_path, line_number,
					"parameter '$param_name' unknown"))

			elseif isa(err, ArgumentError)
				throw(LoadError(gen_config_file_path, line_number,
					"invalid value for '$param_name': $value"))
			else
				throw(err)
			end
		end
	end
	
	return nothing
end

File: src/python/modules/parameters/load_general_configuration.py
from .entities import ParameterData

def parse(value: str) -> str:
    return value


def load_general_configuration(gen_config_file_path: str, params:ParameterData) -> None:
    # Create a dictionary with fields and their types.
    param_names_types = {
        name: typ for name, typ in zip(params.__dataclass_fields__.keys(), params.__dataclass_fields__.values())
    }

    param_given = {name: False for name in param_names_types.keys()}

    lines = []
    try:
        with open(gen_config_file_path, 'r') as file:
            lines = file.readlines()
    except Exception as err:
        raise RuntimeError(f"Cannot read '{gen_config_file_path}'") from err

    if len(lines) == 0:
        raise RuntimeError(f"Cannot read '{gen_config_file_path}'")

    for line_number, line in enumerate(lines, start=1):
        line = line.strip()
        if len(line) == 0 or line.startswith('#'):
            continue

        try:
            param_name, value = line.split()
            if param_name in param_names_types:
                field_type = param_names_types[param_name].type
                if field_type is bool:
                    parsed_value = value.lower() in ("true", "1", "yes", "on")
                elif field_type is str:
                    parsed_value = parse(value)
                else:
                    parsed_value = field_type(value)
                setattr(params, param_name, parsed_value)
                param_given[param_name] = True
                print(f"Parameter '{param_name}' with value '{value}' successfully set as {parsed_value} of type {type(parsed_value)}")
            else:
                raise KeyError(f"Parameter '{param_name}' unknown")

        except ValueError:
            raise RuntimeError(
                f"Error on line {line_number} of '{gen_config_file_path}': missing parameter or value"
            )
        except KeyError as err:
            raise RuntimeError(
                f"Error on line {line_number} of '{gen_config_file_path}': {err}"
            )
        except Exception as err:
            raise RuntimeError(
                f"Error on line {line_number} of '{gen_config_file_path}': invalid value for '{param_name}': {value}"
            ) from err

    return None